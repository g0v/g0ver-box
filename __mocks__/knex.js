import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
