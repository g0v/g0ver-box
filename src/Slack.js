import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
