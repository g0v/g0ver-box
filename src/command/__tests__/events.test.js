/* eslint import/imports-first: 0 */

jest.mock('../../../knexfile', () => ({
  client: 'pg',
  connection: {
    host: '127.0.0.1',
    user: 'postgres',
    password: null,
    database: 'g0ver_projects',
  },
}));

import { client } from 'knex';
import Slack from '../../Slack';
import db from '../../model/Database';
import index from '../';

describe('projects command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'events' };

  it('successful query', async () => {
    client.mockReturnValueOnce([{
      id: 10,
      title: '基礎建設松',
      userId: 'U03B2AB13',
      datetime: '2017-11-05T06:00:00',
      archive: {
        url: 'https://g0v.tw/',
      },
    }, {
      id: 15,
      title: '新聞新聞松',
      userId: 'U03B2AB00',
      datetime: '2017-12-10T14:30:00',
    }]);

    await index(data);
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(1);
    expect(Slack.postMessage.mock.calls).toMatchSnapshot();
  });

  describe('database', () => {
    beforeAll(async () => {
      await db('pg_tables').select('tablename').where('schemaname', 'public').map(({ tablename }) => (
        db.schema.dropTable(tablename)
      ));
      await db.migrate.latest();
    });

    afterAll(async () => {
      await db.destroy();
    });

    it('jothon event', async () => {
      await index({ ...data, text: 'jothon 基礎建設松' });
      await index({ ...data, text: '12-31 23:00' });
      await index({ ...data, text: 'https://events.g0v.tw/' });
      await index({ ...data, text: 'yes' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, jothon 基礎建設松',
      });
    });

    it('events', async () => {
      await index(data);
      expect(Slack.postMessage.mock.calls).toMatchSnapshot();
    });

    it('cancel event', async () => {
      await index({ ...data, text: 'cancel 基礎建設松' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, 取消 基礎建設松 活動',
      });
    });
  });
});
