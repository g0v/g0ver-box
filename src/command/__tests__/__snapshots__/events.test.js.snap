// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projects command database events 1`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "<!date^1514732400000^{date_num} at {time}|unknown>
聯絡人: <@U03B2AB13>",
          "title": "基礎建設松",
          "title_link": "https://events.g0v.tw/",
        },
      ],
      "channel": "D100",
      "text": "下列活動歡迎你的參加：",
    },
  ],
]
`;

exports[`projects command successful query 1`] = `
Array [
  "select * from event where deleted_at is null limit 1000",
]
`;

exports[`projects command successful query 2`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "<!date^1509861600000^{date_num} at {time}|unknown>
聯絡人: <@U03B2AB13>",
          "title": "基礎建設松",
          "title_link": "https://g0v.tw/",
        },
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "<!date^1512916200000^{date_num} at {time}|unknown>
聯絡人: <@U03B2AB00>",
          "title": "新聞新聞松",
          "title_link": null,
        },
      ],
      "channel": "D100",
      "text": "下列活動歡迎你的參加：",
    },
  ],
]
`;
